x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java

```diff
@@ -43,60 +50,75 @@ public class SearchApplicationTemplate implements ToXContentObject, Writeable {
+        null
     );
+
     private final Script script;
+    private final TemplateParamValidator templateParamValidator;
 
     public SearchApplicationTemplate(StreamInput in) throws IOException {
         this.script = in.readOptionalWriteable(Script::new);
+        this.templateParamValidator = in.readOptionalWriteable(TemplateParamValidator::new);
     }
 
-    public SearchApplicationTemplate(Script script) {
-        if (script != null && script.getLang() != null) {
-            if (MustacheScriptEngine.NAME.equals(script.getLang()) == false) {
-                throw new IllegalArgumentException("only [" + MustacheScriptEngine.NAME + "] scripting language is supported");
-            }
-        }
-        if (script != null && script.getType().getId() != ScriptType.INLINE.getId()) {
-            throw new IllegalArgumentException("only [" + ScriptType.INLINE.getName() + "] script type is supported");
-        }
-
-        this.script = script;
-    }
+    private static final ConstructingObjectParser<SearchApplicationTemplate, Void> PARSER = new ConstructingObjectParser<>(
+        "search_template",
+        p -> new SearchApplicationTemplate((Script) p[0], (TemplateParamValidator) p[1])
+    );
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
-        builder.field(SearchApplication.TEMPLATE_SCRIPT_FIELD.getPreferredName(), script);
+        if (script != null) {
+            builder.field(TEMPLATE_SCRIPT_FIELD.getPreferredName(), script);
+        }
+        if (templateParamValidator != null) {
+            builder.field(DICTIONARY_FIELD.getPreferredName(), templateParamValidator);
+        }
         builder.endObject();
         return builder;
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeOptionalWriteable(script);
+        out.writeOptionalWriteable(templateParamValidator);
     }
 
     public static SearchApplicationTemplate parse(XContentParser parser) {
         return PARSER.apply(parser, null);
     }
 
-    private static final ConstructingObjectParser<SearchApplicationTemplate, Void> PARSER = new ConstructingObjectParser<>(
-        "search_template",
-        p -> new SearchApplicationTemplate((Script) p[0])
-    );
-
     static {
         PARSER.declareObject(
             optionalConstructorArg(),
             (p, c) -> Script.parse(p, Script.DEFAULT_TEMPLATE_LANG),
             SearchApplication.TEMPLATE_SCRIPT_FIELD
         );
+        PARSER.declareObject(optionalConstructorArg(), (p, c) -> {
+            try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+                return new TemplateParamValidator(builder.copyCurrentStructure(p));
+            }
+        }, DICTIONARY_FIELD);
+    }
+
+    public SearchApplicationTemplate(Script script, @Nullable TemplateParamValidator templateParamValidator) {
```

Comment:
Why is `templateParamValidator` nullable?

Reply:
In case there is no `dictionary`, there's no validation. That is what having a `null` `TemplateParamValidator` means.

Reply:
Got it. I assume we don't want to take a hard line and require a dictionary?

Reply:
I can see dictionary being added later in the game, when people is already comfortable with the query they are producing and then want to harden the parameters passed in. I think it stands in the way otherwise.

Reply:
OK, that makes sense. Thanks for discussing it!